src/fetch/webfinger.rs

```diff
@@ -89,22 +110,39 @@ where
 /// # Ok::<(), anyhow::Error>(())
 /// }).unwrap();
 ///```
-pub fn extract_webfinger_name<T>(query: &str, data: &Data<T>) -> Result<String, Error>
+pub fn extract_webfinger_name<'i, T>(query: &'i str, data: &Data<T>) -> Result<&'i str, Error>
 where
     T: Clone,
 {
-    // Regex to extract usernames from webfinger query. Supports different alphabets using `\p{L}`.
-    // TODO: would be nice if we could implement this without regex and remove the dependency
-    let result = Regex::new(&format!(r"^acct:([\p{{L}}0-9_]+)@{}$", data.domain()))
-        .map_err(|_| Error::WebfingerRegexFailed)
-        .and_then(|regex| {
-            regex
-                .captures(query)
-                .and_then(|c| c.get(1))
-                .ok_or_else(|| Error::WebfingerRegexFailed)
-        })?;
+    fn accept_char_name(c: char) -> bool {
+        use unicode_properties::{GeneralCategoryGroup, UnicodeGeneralCategory};
+
+        c == '_' || c.is_digit(10) || c.general_category_group() == GeneralCategoryGroup::Letter
+    }
+
+    let Some(rem) = query.strip_prefix("acct:") else {
+        return Err(WebFingerError::WrongFormat.to_crate_error());
+    };
+
+    let mut rem_iter = rem.char_indices();
+    while let Some((idx, c)) = rem_iter.next() {
+        if c == '@' {
+            assert_eq!(c.len_utf8(), 1);
+            if idx == 0 {
+                return Err(WebFingerError::WrongFormat.into());
+            }
+
+            if &rem[idx + 1..] != data.domain() {
+                return Err(WebFingerError::WrongDomain.into());
+            }
+
+            return Ok(&rem[..idx]);
+        } else if !accept_char_name(c) {
+            return Err(WebFingerError::WrongFormat.into());
+        }
```

Comment:
Ah so you managed to get rid of regex to validate the webfinger query. Then you should be able to get rid of the regex dependency. Also please add some comments to explain what the code is doing.

Reply:
I added some documentation.

The `regex` crate is also used to [parse the `Signature` header for the `keyId` of the sender](https://github.com/LemmyNet/activitypub-federation-rust/blob/main/src/http_signatures.rs#L164). I looked at actually parsing that but it seems that doing it properly would be a lot more involved. The [`sfv`](https://docs.rs/sfv) crate has in theory the correct parser but it enforces the rule from RFC 8941 that `keyId` should not be allowed to have a capital letter, which is visibly not respected by ActivityPub...

Reply:
I see, in that case it seems much simpler to keep the regex here and capture the domain for checking `WrongDomain` error.

Reply:
I have looked a bit more at the spec because I was curious to know where the `Letter` category in the regex came from (I initially only copied the behaviour of the regex).

It looks like the correct way to parse the `acct` part of a webfinger url would be to parse it as a `Url`, where `acct` is a scheme (like https), defined in https://datatracker.ietf.org/doc/html/rfc7565

It also looks like the webfinger endpoints can accept other versions of the url for an account. For example:

```
curl https://pouet.chapril.org/.well-known/webfinger?resource=acct:sgued@pouet.chapril.org
curl https://pouet.chapril.org/.well-known/webfinger?resource=https://pouet.chapril.org/users/sgued
curl https://pouet.chapril.org/.well-known/webfinger?resource=https://pouet.chapril.org/@sgued
```

All return the same value, but on the lemmy side, `curl https://lemmy.ml/.well-known/webfinger?resource=https://lemmy.ml/u/nutomic` returns an error.

This should probably be done in another PR though. I'll restore the `Regex` implementation so that this PR can be merged.

Reply:
Yes please revert it. The current implementation is working, so Id rather keep it as is and avoid any risk of breakage. Especially when there is no practical benefit from the change.